### PR TITLE
Preliminary Kotlin Support

### PR DESCRIPTION
**Pull request by kageiit**
Ref #400

This adds basic support only for building and testing pure kotlin modules. `kotlinc` is invoked out of memory and needs to be improved on buck's side to make it faster in the future. See #323

Generated kotlin rules look like

```
kotlin_library(
	name = 'src_main',
	srcs = glob([
		'src/main/kotlin/**/*.kt',
	]),
	tests = [
		':test_main',
	],
	source = '7',
	target = '7',
	deps = [
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-runtime-1.0.6.jar',
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-stdlib-1.0.6.jar',
	],
	visibility = [
		'PUBLIC',
	],
)

kotlin_test(
	name = 'test_main',
	srcs = glob([
		'src/test/kotlin/**/*.kt',
	]),
	source = '7',
	target = '7',
	labels = [
		'unit',
		'kotlin',
	],
	deps = [
		'//.okbuck/cache:junit.junit-4.12.jar',
		'//.okbuck/cache:org.hamcrest.hamcrest-core-1.3.jar',
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-runtime-1.0.6.jar',
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-stdlib-1.0.6.jar',
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-test-1.0.6.jar',
		'//.okbuck/cache:org.jetbrains.kotlin.kotlin-test-junit-1.0.6.jar',
		':src_main',
	],
	visibility = [
		'PUBLIC',
	],
)
```

### CODE REVIEW

File: buildSrc/src/main/java/com/uber/okbuck/core/util/KotlinUtil.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                ImmutableMap.of("template-kotlin-version", kotlinVersion));
+        kotlinc.setExecutable(true);
+
+        String kotlinCompiler = rootProject.relativePath(kotlinc);
+        String kotlinRuntime = String.format("%s/lib/%s-%s.jar", KOTLIN_HOME_LOCATION, KOTLIN_RUNTIME_MODULE,
+                kotlinVersion);
+        return Pair.of(kotlinCompiler, kotlinRuntime);
+    }
+
+    private static boolean findKotlin(ResolvedArtifact artifact) {
```
Comment (raviagarwal7) [inline buildSrc/src/main/java/com/uber/okbuck/core/util/KotlinUtil.java]: nit -- probably inline this

**Comment by raviagarwal7** [comment]
Looks good! 
please add the generated kotlin buck rules for the fixture library in the summary for reference.

**Comment by runningcode** [comment]
By "pure kotlin modules" do you mean kotlin modules without annotation processors or kotlin modules which do not include java source files?

**Comment by kageiit** [comment]
> By "pure kotlin modules" do you mean kotlin modules without annotation processors or kotlin modules which do not include java source files?

See #400 

basically support now is only for a module that has `src/main/kotlin`. If you have java sources, you can make a java module and have your kotlin module depend on that (or vice versa). Mixed java/kotlin compilation is not supported by buck yet

**Comment by kuno** [comment]
I am  playing okbuck 0.19.3 with kotlin 1.1.1 on an experimental project, which only contains an app module written in pure kotlin.

The module itself is only have a custom `Application` class and a default activity.

I can build it, installed on devices. But when running, it crashed caused by not found the application class and the default activity class in any path.

Is this the normal behavior at this age ?

**Pull request merged**